File: chaincode/sensorchain/src/firmware.ts
/*
 * Secure Firmware Upgrade contract (production chaincode).
 * Mirrors sensorchain/contracts/firmware.py.
 *
 * Chain of trust: manufacturer signing key registered on-chain →
 * releases published with a verified ECDSA signature → city SPV
 * approval gate → device-side hash verification before flashing →
 * post-flash attestation with auto-flagging on mismatch.
 */

import { createHash, createVerify } from "crypto";
import { Context, Contract, Info, Returns, Transaction } from "fabric-contract-api";
import { deviceKey, DeviceRecord } from "./deviceRegistry";

export interface FirmwareRelease {
  manufacturer: string;
  model: string;
  version: string;
  firmware_hash: string;
  signature: string;
  approved: boolean;
  approved_by?: string;
  published_at: number;
}

export function firmwareSigningPayload(
  manufacturer: string, model: string, version: string, firmwareHash: string,
): Buffer {
  return Buffer.from(`firmware|${manufacturer}|${model}|${version}|${firmwareHash}`, "utf8");
}

@Info({
  title: "Firmware",
  description: "Manufacturer-signed firmware releases with approval gate and rollout attestation",
})
export class FirmwareContract extends Contract {
  constructor() {
    super("Firmware");
  }

  private makerKey(ctx: Context, name: string): string {
    return ctx.stub.createCompositeKey("manufacturer", [name]);
  }

  private releaseKey(ctx: Context, model: string, version: string): string {
    return ctx.stub.createCompositeKey("firmware", [model, version]);
  }

  @Transaction()
  @Returns("string")
  public async RegisterManufacturer(ctx: Context, name: string, publicKeyPem: string): Promise<string> {
    const key = this.makerKey(ctx, name);
    const existing = await ctx.stub.getState(key);
    if (existing.length > 0) {
      throw new Error(`manufacturer ${name} already registered`);
    }
    const record = {
      name,
      public_key_pem: publicKeyPem,
      key_fingerprint: createHash("sha256").update(publicKeyPem).digest("hex").slice(0, 16),
      registered_at: ctx.stub.getTxTimestamp().seconds.toNumber(),
    };
    await ctx.stub.putState(key, Buffer.from(JSON.stringify(record)));
    ctx.stub.setEvent("ManufacturerRegistered", Buffer.from(JSON.stringify({
      name, key_fingerprint: record.key_fingerprint,
    })));
    return JSON.stringify(record);
  }

  @Transaction()
  @Returns("string")
  public async PublishFirmware(
    ctx: Context, manufacturer: string, model: string,
    version: string, firmwareHash: string, signatureHex: string,
  ): Promise<string> {
    const makerBytes = await ctx.stub.getState(this.makerKey(ctx, manufacturer));
    if (makerBytes.length === 0) {
      throw new Error(`unknown manufacturer ${manufacturer}`);
    }
    const maker = JSON.parse(makerBytes.toString()) as { public_key_pem: string };
    const verifier = createVerify("SHA256");
    verifier.update(firmwareSigningPayload(manufacturer, model, version, firmwareHash));
    if (!verifier.verify(maker.public_key_pem, Buffer.from(signatureHex, "hex"))) {
      ctx.stub.setEvent("FirmwareSignatureInvalid", Buffer.from(JSON.stringify({
        manufacturer, model, version,
      })));
      throw new Error(
        `firmware ${model} ${version} rejected: signature does not match ${manufacturer}'s registered signing key`);
    }
    const key = this.releaseKey(ctx, model, version);
    const existing = await ctx.stub.getState(key);
    if (existing.length > 0) {
      throw new Error(`firmware ${model} ${version} already published`);
    }
    const release: FirmwareRelease = {
      manufacturer, model, version,
      firmware_hash: firmwareHash,
      signature: signatureHex,
      approved: false,
      published_at: ctx.stub.getTxTimestamp().seconds.toNumber(),
    };
    await ctx.stub.putState(key, Buffer.from(JSON.stringify(release)));
    ctx.stub.setEvent("FirmwarePublished", Buffer.from(JSON.stringify({
      manufacturer, model, version, firmware_hash: firmwareHash,
    })));
    return JSON.stringify(release);
  }

  @Transaction()
  @Returns("string")
  public async ApproveFirmware(ctx: Context, model: string, version: string, approver: string): Promise<string> {
    const key = this.releaseKey(ctx, model, version);
    const bytes = await ctx.stub.getState(key);
    if (bytes.length === 0) {
      throw new Error(`no published firmware ${model} ${version}`);
    }
    const release = JSON.parse(bytes.toString()) as FirmwareRelease;
    release.approved = true;
    release.approved_by = approver;
    await ctx.stub.putState(key, Buffer.from(JSON.stringify(release)));
    ctx.stub.setEvent("FirmwareApproved", Buffer.from(JSON.stringify({ model, version, approved_by: approver })));
    return JSON.stringify(release);
  }

  @Transaction(false)
  @Returns("string")
  public async VerifyImage(ctx: Context, model: string, version: string, imageHash: string): Promise<string> {
    const bytes = await ctx.stub.getState(this.releaseKey(ctx, model, version));
    if (bytes.length === 0) {
      return JSON.stringify({ ok: false, reason: "unknown_release" });
    }
    const release = JSON.parse(bytes.toString()) as FirmwareRelease;
    if (!release.approved) {
      return JSON.stringify({ ok: false, reason: "not_approved_for_rollout" });
    }
    if (release.firmware_hash !== imageHash) {
      return JSON.stringify({
        ok: false, reason: "image_hash_mismatch",
        expected: release.firmware_hash, got: imageHash,
      });
    }
    return JSON.stringify({ ok: true, release });
  }

  @Transaction()
  @Returns("string")
  public async ReportUpdate(
    ctx: Context, deviceId: string, model: string, version: string, installedHash: string,
  ): Promise<string> {
    const devKey = deviceKey(ctx, deviceId);
    const devBytes = await ctx.stub.getState(devKey);
    if (devBytes.length === 0) {
      throw new Error(`unknown device ${deviceId}`);
    }
    const relBytes = await ctx.stub.getState(this.releaseKey(ctx, model, version));
    if (relBytes.length === 0) {
      throw new Error(`no published firmware ${model} ${version}`);
    }
    const device = JSON.parse(devBytes.toString()) as DeviceRecord & Record<string, unknown>;
    const release = JSON.parse(relBytes.toString()) as FirmwareRelease;

    if (release.approved && installedHash === release.firmware_hash) {
      device.firmware_model = model;
      device.firmware_version = version;
      await ctx.stub.putState(devKey, Buffer.from(JSON.stringify(device)));
      ctx.stub.setEvent("FirmwareUpdateApplied", Buffer.from(JSON.stringify({
        device_id: deviceId, model, version,
      })));
      return JSON.stringify({ ok: true, device });
    }

    const reason = !release.approved ? "unapproved_release" : "installed_hash_mismatch";
    device.status = "flagged";
    device.status_reason = `firmware violation: ${reason}`;
    await ctx.stub.putState(devKey, Buffer.from(JSON.stringify(device)));
    ctx.stub.setEvent("FirmwareHashMismatch", Buffer.from(JSON.stringify({
      device_id: deviceId, model, version, reason,
      expected: release.firmware_hash, got: installedHash,
    })));
    return JSON.stringify({ ok: false, reason, device });
  }
}

File: sensorchain/contracts/firmware.py
"""Secure Firmware Upgrade contract.

Closes the "secure upgrade mechanisms" requirement of the IoT/IIoT use
case. The trust chain:

  1. A manufacturer is registered on-chain with its firmware-signing
     public key.
  2. Each firmware release (model, version, SHA-256 of the image) is
     published with the manufacturer's signature; the contract verifies
     it before anchoring. Unsigned or wrongly-signed releases are
     rejected.
  3. The city SPV approves a published release for rollout. Devices
     only ever apply approved firmware.
  4. Before flashing, a device (or its gateway) fetches the on-chain
     record and checks the image hash against it. After flashing it
     reports the installed hash; a mismatch auto-flags the device and
     emits a FirmwareHashMismatch event — a tampered or trojaned image
     is caught at rollout time, on an auditable channel.
"""

from __future__ import annotations

import time

from ..identity import key_fingerprint, verify_signature
from ..ledger import Ledger


def firmware_signing_payload(manufacturer: str, model: str, version: str, firmware_hash: str) -> bytes:
    """The exact bytes a manufacturer signs when releasing firmware."""
    return f"firmware|{manufacturer}|{model}|{version}|{firmware_hash}".encode("utf-8")


class FirmwareContract:
    CONTRACT = "firmware"

    def __init__(self, ledger: Ledger):
        self.ledger = ledger

    # ------------------------------------------------------------------
    # Manufacturer signing keys
    # ------------------------------------------------------------------
    def register_manufacturer(self, name: str, public_key_pem: str) -> dict:
        key = f"manufacturer~{name}"
        if self.ledger.get_state(key) is not None:
            raise ValueError(f"manufacturer {name} already registered")
        record = {
            "name": name,
            "public_key_pem": public_key_pem,
            "key_fingerprint": key_fingerprint(public_key_pem),
            "registered_at": time.time(),
        }
        self.ledger.put_state(self.CONTRACT, "registerManufacturer", key, record)
        self.ledger.emit_event("ManufacturerRegistered", {
            "name": name, "key_fingerprint": record["key_fingerprint"],
        })
        return record

    # ------------------------------------------------------------------
    # Release lifecycle: publish (manufacturer) → approve (city SPV)
    # ------------------------------------------------------------------
    @staticmethod
    def _release_key(model: str, version: str) -> str:
        return f"firmware~{model}~{version}"

    def publish_firmware(self, manufacturer: str, model: str, version: str,
                         firmware_hash: str, signature_hex: str) -> dict:
        maker = self.ledger.get_state(f"manufacturer~{manufacturer}")
        if maker is None:
            raise KeyError(f"unknown manufacturer {manufacturer}")
        payload = firmware_signing_payload(manufacturer, model, version, firmware_hash)
        if not verify_signature(maker["public_key_pem"], payload, signature_hex):
            self.ledger.emit_event("FirmwareSignatureInvalid", {
                "manufacturer": manufacturer, "model": model, "version": version,
            })
            raise PermissionError(
                f"firmware {model} {version} rejected: signature does not match "
                f"{manufacturer}'s registered signing key")
        key = self._release_key(model, version)
        if self.ledger.get_state(key) is not None:
            raise ValueError(f"firmware {model} {version} already published")
        release = {
            "manufacturer": manufacturer,
            "model": model,
            "version": version,
            "firmware_hash": firmware_hash,
            "signature": signature_hex,
            "approved": False,
            "published_at": time.time(),
        }
        self.ledger.put_state(self.CONTRACT, "publishFirmware", key, release)
        self.ledger.emit_event("FirmwarePublished", {
            "manufacturer": manufacturer, "model": model, "version": version,
            "firmware_hash": firmware_hash,
        })
        return release

    def approve_firmware(self, model: str, version: str, approver: str) -> dict:
        release = self.ledger.get_state(self._release_key(model, version))
        if release is None:
            raise KeyError(f"no published firmware {model} {version}")
        updated = dict(release, approved=True, approved_by=approver, approved_at=time.time())
        self.ledger.put_state(self.CONTRACT, "approveFirmware", self._release_key(model, version), updated)
        self.ledger.emit_event("FirmwareApproved", {
            "model": model, "version": version, "approved_by": approver,
        })
        return updated

    def get_release(self, model: str, version: str) -> dict | None:
        return self.ledger.get_state(self._release_key(model, version))

    def latest_approved(self, model: str) -> dict | None:
        releases = [v for _, v in self.ledger.query_by_prefix(f"firmware~{model}~") if v["approved"]]
        return max(releases, key=lambda r: r["published_at"]) if releases else None

    # ------------------------------------------------------------------
    # Device side: verify before flashing, report after
    # ------------------------------------------------------------------
    def verify_image(self, model: str, version: str, image_hash: str) -> dict:
        """What a device/gateway calls BEFORE flashing an image it received."""
        release = self.get_release(model, version)
        if release is None:
            return {"ok": False, "reason": "unknown_release"}
        if not release["approved"]:
            return {"ok": False, "reason": "not_approved_for_rollout"}
        if release["firmware_hash"] != image_hash:
            return {"ok": False, "reason": "image_hash_mismatch",
                    "expected": release["firmware_hash"], "got": image_hash}
        return {"ok": True, "release": release}

    def report_update(self, device_id: str, model: str, version: str, installed_hash: str) -> dict:
        """Post-flash attestation. A hash mismatch flags the device on-chain."""
        device_key = f"device~{device_id}"
        device = self.ledger.get_state(device_key)
        if device is None:
            raise KeyError(f"unknown device {device_id}")
        release = self.get_release(model, version)
        if release is None:
            raise KeyError(f"no published firmware {model} {version}")

        if release["approved"] and installed_hash == release["firmware_hash"]:
            updated = dict(device, firmware_model=model, firmware_version=version,
                           firmware_updated_at=time.time())
            self.ledger.put_state(self.CONTRACT, "reportUpdate", device_key, updated)
            self.ledger.emit_event("FirmwareUpdateApplied", {
                "device_id": device_id, "model": model, "version": version,
            })
            return {"ok": True, "device": updated}

        reason = "unapproved_release" if not release["approved"] else "installed_hash_mismatch"
        flagged = dict(device, status="flagged",
                       status_reason=f"firmware violation: {reason}",
                       status_updated_at=time.time())
        self.ledger.put_state(self.CONTRACT, "reportUpdate", device_key, flagged)
        self.ledger.emit_event("FirmwareHashMismatch", {
            "device_id": device_id, "model": model, "version": version,
            "reason": reason,
            "expected": release["firmware_hash"], "got": installed_hash,
        })
        return {"ok": False, "reason": reason, "device": flagged}
